src/ds/redblacktree.h

```diff
@@ -11,16 +11,40 @@
 namespace snmalloc
 {
 #ifdef __cpp_concepts
+  /**
+   * The representation must define two types.  `Contents` defines some
+   * identifier that can be mapped to a node as a value type.  `Holder` defines
```

Comment:
With `Holder*` and `Holder&`, you required the holders to be things that were allocated and had a lifetime that lasted beyond the lifetime of the pointers / references to them, which didn't allow them to be wrapper types.  With the new abstraction, `Holder` (which, I agree, is a bad name.  I prefer `Handle` because `Reference` implies that I can't set the thing that it refers to), the assumption that the storage location and the contents types are the same is in precisely one place, and that's easy to remove (done in the latest revision).

Reply:
The previous solution allowed the storage location to have a more interesting type that could control the access.  This is different to perhaps the more standard smart pointer approach.  The original example in redblack.cc did perform interesting manipulations that related the stored value to the storage location.  I think your approach is less invasive on the storage locations type by using smartptr type instead.  It does require an additional type `decltype(root)`, which I think of as being `Holder` before this PR.

I think it would be good to switch `Holder` to `Handle` as I don't think `Holder` is the correct description now you have changed its meaning.